giftless/storage/aws_s3.py

```diff
@@ -0,0 +1,110 @@
+import os
+from typing import Any, BinaryIO, Dict, Iterable, Optional
+
+import boto3  # type: ignore
+import botocore  # type: ignore
+
+from giftless.storage import ExternalStorage, StreamingStorage
+from giftless.storage.exc import ObjectNotFound
+
+
+class AwsS3Storage(StreamingStorage, ExternalStorage):
+    """AWS S3 Blob Storage backend.
+    """
+
+    def __init__(self, aws_s3_bucket_name: str, path_prefix: Optional[str] = None, **_):
+        self.aws_s3_bucket_name = aws_s3_bucket_name
+        self.path_prefix = path_prefix
+        self.s3 = boto3.resource('s3')
+        self.s3_client = boto3.client('s3')
+
+    def get(self, prefix: str, oid: str) -> Iterable[bytes]:
+        if not self.exists(prefix, oid):
+            raise ObjectNotFound()
+        result: Iterable[bytes] = self._s3_object(prefix, oid).get()['Body']
+        return result
+
+    def put(self, prefix: str, oid: str, data_stream: BinaryIO) -> int:
+        completed = []
+
+        def upload_callback(size):
+            completed.append(size)
+
+        bucket = self.s3.Bucket(self.aws_s3_bucket_name)
+        bucket.upload_fileobj(data_stream, self._get_blob_path(prefix, oid), Callback=upload_callback)
+        return sum(completed)
+
+    def exists(self, prefix: str, oid: str) -> bool:
+        s3_object = self._s3_object(prefix, oid)
+        try:
+            s3_object.load()
```

Comment:
I have learned that there used to exists an `exist` api call in older boto2. It got removed long time ago in `boto3`.
I have found a few SO answers about the issue and there was general consensus that this should be the best way to check if object exists.
The `S3 Object load()` (docs)[https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Object.load] calls `S3 Client head_object()` (docs)[https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.head_object] which fetches only the metadata without the object itself.
In case of `ClientError` I have also added another check for error code `404` to be fully sure that the exception is a result of the object not existing. Not some other resons.

I think it's a good mechanism but please let me know if you feel convinces by my reasoning.

Reply:
I think that it's irrelevant now after using `get_size` to determine object existence with newest changes.